lib/tree.rb

```diff
@@ -16,42 +16,76 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def add(key, value)
-    raise NotImplementedError
+  # Time Complexity: O(log(n))
+  # Space Complexity: O(log(n))
+  def add(key, value = nil, current = @root)
+    @root = TreeNode.new(key, value) and return if @root.nil?
+
+    if key <= current.key
+      current.left = TreeNode.new(key, value) and return if current.left.nil?
+      add(key, value, current.left)
+    else 
+      current.right = TreeNode.new(key, value) and return if current.right.nil?
+      add(key, value, current.right)
+    end
+
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def find(key)
-    raise NotImplementedError
+  # Time Complexity: O(log(n))
+  # Space Complexity: O(log(n))
+  def find(key, current = @root)
```

Comment:
👍 The time/space complexities are correct if the tree is balanced and O(n) otherwise.